src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -3659,8 +3659,387 @@ class StubGenerator: public StubCodeGenerator {
+        0xeada7dd6cde0eb1el, 0xf57d4f7fee6ed178l, 0x06f067aa72176fbal,
+        0x0a637dc5a2c898a6l, 0x113f9804bef90dael, 0x1b710b35131c471bl,
+        0x28db77f523047d84l, 0x32caab7b40c72493l, 0x3c9ebe0a15c9bebcl,
+        0x431d67c49c100d4cl, 0x4cc5d4becb3e42b6l, 0x597f299cfc657e2al,
+        0x5fcb6fab3ad6faecl, 0x6c44198c4a475817l
+      };
+      const int const_add = vset_sew == Assembler::e32 ? 16 : 32;
+
+      __ align(CodeEntryAlignment);
+      StubCodeMark mark(_cgen, "StubRoutines", stub_name(vset_sew, multi_block));
+      address start = __ pc();
+
+      Register buf   = c_rarg0;
+      Register state = c_rarg1;
+      Register ofs   = c_rarg2;
+      Register limit = c_rarg3;
+      Register consts =  t2; // caller saved
+      Register state_c = x28; // caller saved
+      VectorRegister vindex = v1;
+
+      Label multi_block_loop;
+
+      __ enter();
+
+      address constant_table = vset_sew == Assembler::e32 ? (address)round_consts_256 : (address)round_consts_512;
+      la(consts, ExternalAddress(constant_table));
+
+      // Register use in this function:
+      //
+      // VECTORS
+      //  v10 - v13 (512/1024-bits / 4*128/256 bits / 4*4*32/65 bits), hold the message
+      //             schedule words (Wt). They start with the message block
+      //             content (W0 to W15), then further words in the message
+      //             schedule generated via vsha2ms from previous Wt.
+      //   Initially:
+      //     v10 = W[  3:0] = { W3,  W2,  W1,  W0}
+      //     v11 = W[  7:4] = { W7,  W6,  W5,  W4}
+      //     v12 = W[ 11:8] = {W11, W10,  W9,  W8}
+      //     v13 = W[15:12] = {W15, W14, W13, W12}
+      //
+      //  v16 - v17 hold the working state variables (a, b, ..., h)
+      //    v16 = {f[t],e[t],b[t],a[t]}
+      //    v17 = {h[t],g[t],d[t],c[t]}
+      //   Initially:
+      //    v16 = {H5i-1, H4i-1, H1i-1 , H0i-1}
+      //    v17 = {H7i-i, H6i-1, H3i-1 , H2i-1}
+      //
+      //  v0 = masks for vrgather/vmerge. Single value during the 16 rounds.
+      //
+      //  v14 = temporary, Wt+Kt
+      //  v15 = temporary, Kt
+      //
+      //  v18/v19 = temporaries, in the epilogue, to re-arrange
+      //            and byte-swap v16/v17
+      //
+      //  v26/v27 = hold the initial values of the hash, byte-swapped.
+      //
+      //  v30/v31 = used to generate masks, vrgather indices.
+      //
+      // During most of the function the vector state is configured so that each
+      // vector is interpreted as containing four 32/64 bits (e32/e64) elements (128/256 bits).
+
+      // Set vectors as 4 * 32/64 bits
+      //
+      // e32/e64: vector of 32b/64b/4B/8B elements
+      // m1: LMUL=1
+      // ta: tail agnostic (don't care about those lanes)
+      // ma: mask agnostic (don't care about those lanes)
+      // x0 is not written, we known the number of vector elements.
+
+      __ vsetivli(x0, 4, vset_sew, Assembler::m1, Assembler::ma, Assembler::ta);
```

Comment:
The sha instruction don't care about lmul, except only in that lmul*vlen == 256.
So there is no code difference except just setting m2 in case of 128.
These crypto uses 'EGW' element group width which is specified.
As quad rounds dependent on result from previous quad round, you can only do 4xe64(for 256 e32), no more no less.

Yes, good catch! I must remember to tests multiple vlens. Thanks!

Reply:
I suppose we would need some more conditions to fix the register usage for the sha512 case in places like indexed load/store, message block load, etc. Because we will work with register pairs for this case. I am not sure if that's a good way.

Reply:
Is it possible to just use one pieice of code to work for both sha-256/sha-512? I mean without conditions everywhere. As the difference between 256/512 is just some contants and rounds number.

Reply:
Yes, we can't use v10 and v11, as there is only two alternative.
vlen=128 we need to use two registers, and vlen>=256 we need one.
Simple issue, fixed.